src/Symfony/Component/Serializer/Normalizer/DataUriNormalizer.php

```diff
@@ -0,0 +1,157 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Serializer\Normalizer;
+
+use Symfony\Component\HttpFoundation\File\File;
+use Symfony\Component\HttpFoundation\File\MimeType\MimeTypeGuesser;
+use Symfony\Component\HttpFoundation\File\MimeType\MimeTypeGuesserInterface;
+use Symfony\Component\Serializer\Exception\InvalidArgumentException;
+use Symfony\Component\Serializer\Exception\UnexpectedValueException;
+
+/**
+ * Normalizes an {@see \SplFileInfo} object to a data URI.
+ * Denormalizes a data URI to a {@see \SplFileObject} object.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+class DataUriNormalizer implements NormalizerInterface, DenormalizerInterface
+{
+    /**
+     * @var MimeTypeGuesserInterface
+     */
+    private $mimeTypeGuesser;
+
+    public function __construct(MimeTypeGuesserInterface $mimeTypeGuesser = null)
+    {
+        if (null === $mimeTypeGuesser && class_exists('Symfony\Component\HttpFoundation\File\MimeType\MimeTypeGuesser')) {
+            $mimeTypeGuesser = MimeTypeGuesser::getInstance();
+        }
+
+        $this->mimeTypeGuesser = $mimeTypeGuesser;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function normalize($object, $format = null, array $context = array())
+    {
+        if (!$object instanceof \SplFileInfo) {
+            throw new InvalidArgumentException('The object must be an instance of "\SplFileInfo".');
+        }
+
+        $mimeType = $this->getMimeType($object);
+        $splFileObject = $this->extractSplFileObject($object);
+
+        $data = '';
+
+        $splFileObject->rewind();
+        while (!$splFileObject->eof()) {
+            $data .= $splFileObject->fgets();
+        }
+
+        if ('text' === explode('/', $mimeType, 2)[0]) {
+            return sprintf('data:%s,%s', $mimeType, rawurlencode($data));
+        }
+
+        return sprintf('data:%s;base64,%s', $mimeType, base64_encode($data));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supportsNormalization($data, $format = null)
+    {
+        return $data instanceof \SplFileInfo;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * Regex adapted from Brian Grinstead code.
+     *
+     * @see https://gist.github.com/bgrins/6194623
+     *
+     * @throws InvalidArgumentException
+     * @throws UnexpectedValueException
+     */
+    public function denormalize($data, $class, $format = null, array $context = array())
+    {
+        if (!preg_match('/^data:([a-z0-9]+\/[a-z0-9]+(;[a-z0-9\-]+\=[a-z0-9\-]+)?)?(;base64)?,[a-z0-9\!\$\&\\\'\,\(\)\*\+\,\;\=\-\.\_\~\:\@\/\?\%\s]*\s*$/i', $data)) {
```

Comment:
@dunglas the original RegEx does not match `application/octet-stream` because the second part of the mime type is not allowed to contain a dash.

Which means the DataUriNormalizer can't match an `data uri` that it produced (when no mime type info was available).

Also there's an issue in the `normalizer` logic, the Symfony mime type guesser tries to fetch the path of the file and assert that the file exists. This does not work with `php://memory` / `php://temp` or `data:` type files.

I could submit a PR with the aforementioned speedups as well as some logic to make the mime type guessing part optional (ex.: supply a mime type via `$context`).